### PR TITLE
Improve BLE shutdown.

### PR DESCRIPTION
**Pull request by floitsch**
Improves the shutdown if the BLE isn't shut down nicely from the Toit side.

### CODE REVIEW

File: src/resources/ble_esp32.cc [hunk 1]
```diff
@@ -460,8 +458,13 @@ class BleServiceResource:
...
 
+  void remove_characteristic(BleCharacteristicResource* characteristic) {
+    characteristics_.unlink(characteristic);
+  }
+
   void clear_characteristics() {
-    while (auto characteristic = characteristics_.remove_first()) {
+    while (!characteristics_.is_empty()) {
+      auto characteristic = characteristics_.first();
       group()->unregister_resource(characteristic);
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:468]: Maybe add a comment that explains that unregistering will remove the element from linked list?

File: src/resources/ble_esp32.cc [hunk 2]
```diff
@@ -575,12 +578,24 @@ class ServiceContainer : public BleErrorCapableResource {
   ServiceContainer(BleResourceGroup* group, Kind kind)
       : BleErrorCapableResource(group, kind) {}
 
+  ~ServiceContainer() override {
+    while (!services_.is_empty()) {
+      auto service = services_.first();
+      group()->unregister_resource(service);
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:584]: Maybe add a comment that explains that unregistering will remove the element from linked list?

File: src/resources/ble_esp32.cc [hunk 3]
```diff
@@ -575,12 +578,24 @@ class ServiceContainer : public BleErrorCapableResource {
...
+
   virtual T* type() = 0;
   BleServiceResource* get_or_create_service_resource(ble_uuid_any_t uuid, uint16 start, uint16 end);
   ServiceResourceList& services() { return services_; }
 
   void clear_services() {
-    while (auto service = services_.remove_first()) {
+    while (!services_.is_empty()) {
+      auto service = services_.first();
       group()->unregister_resource(service);
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:599]: Maybe add a comment that explains that unregistering will remove the element from linked list?

File: src/resources/ble_esp32.cc [hunk 4]
```diff
@@ -647,12 +666,28 @@ class BleRemoteDeviceResource : public ServiceContainer<BleRemoteDeviceResource>
   uint16 handle() const { return handle_; }
   void set_handle(uint16 handle) { handle_ = handle; }
 
+  int connect(uint8 own_addr_type, ble_addr_t* addr) {
+    int err = ble_gap_connect(own_addr_type, addr, 3000, null,
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:670]: What should happen if we're already connected?

File: src/resources/ble_esp32.cc [hunk 5]
```diff
@@ -759,17 +794,43 @@ uint32_t BleResourceGroup::on_event(Resource* resource, word data, uint32_t stat
...
 }
 
+BleCharacteristicResource::~BleCharacteristicResource() {
+  while (!subscriptions_.is_empty()) {
+    auto subscription = subscriptions_.remove_first();
+    delete subscription;
+  }
+  while (!descriptors_.is_empty()) {
+    auto descriptor = descriptors_.first();
+    group()->unregister_resource(descriptor);
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:812]: Maybe add a comment that explains that unregistering will remove the element from linked list?

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/resources/ble_esp32.cc (hunk 1 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc:469]: done.

File: src/resources/ble_esp32.cc (hunk 2 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc:586]: done.

File: src/resources/ble_esp32.cc (hunk 3 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc:602]: done.

File: src/resources/ble_esp32.cc (hunk 4 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc:674]: I added a check that returns `BLE_ERR_SUCCESS` if it is already connected.

File: src/resources/ble_esp32.cc (hunk 5 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc:817]: done.

**Pull request merged**